tools/first_observation_experiment.rb

```diff
@@ -33,136 +157,275 @@ def quality_metric_observation_ids( observation_ids, metrics )
-  elsif !observation.georeferenced? || !observation.observed_on? || ( !observation.photos? && !observation.sounds? ) ||
-      observation.human? || !observation.quality_metrics_pass?
-    if !observation.georeferenced? || !observation.observed_on? || ( !observation.photos? && !observation.sounds? )
-      set = "error_missing"
-      errors.concat(
-        ["georeferenced", "observed_on"].reject {| error | observation.public_send( "#{error}?" ) } +
-        ( observation.photos? || observation.sounds? ? [] : ["media"] )
-      )
-    elsif ["recent", "evidence", "location", "date"].
-        any? {| error | quality_metric_observation_ids( [observation.id], error ).count == 1 }
-      set = "error_dqa"
-      errors.concat(
-        ["recent", "evidence", "location", "date"].
-        select {| error | quality_metric_observation_ids( [observation.id], error ).count == 1 }
-      )
-    elsif quality_metric_observation_ids( [observation.id], "subject" ).count == 1
-      set = "error_subject"
-    elsif observation.human? || quality_metric_observation_ids( [observation.id], "wild" ).count == 1
-      set = "captive_or_human"
-    end
-  elsif !observation.appropriate?
-    set = "copyright"
+    cohort_day = ( ( current_day - current_day_to_iterate ) / ( 60 * 60 * 24 ) ).to_i
+    puts [cohort, cohort_day].join( " " )
+
+    user_ids = cohort_data[cohort].keys.map {| key | key.to_s.to_i }
+    obs_data = get_obs_span( current_day_to_iterate, current_day, user_ids )
+
+    casual = obs_data.
+      select {| _, v | ( v[:needs_id].nil? || v[:needs_id].zero? ) && ( v[:research].nil? || v[:research].zero? ) }.
+      keys
+    needs_id = obs_data.
+      select {| _, v | v[:needs_id]&.positive? && ( v[:research].nil? || v[:research].zero? ) }.
+      keys
+    research = obs_data.select {| _, v | v[:research]&.positive? }.keys
+    no_obs = ( user_ids - casual - needs_id - research )
+
+    captives = get_captives_span( current_day_to_iterate, current_day, casual )
+    error = casual - captives
+
+    no_obs.each {| id | cohort_data[cohort][id.to_s.to_sym]["day#{cohort_day}"] = "no_obs" }
+    error.each {| id | cohort_data[cohort][id.to_s.to_sym]["day#{cohort_day}"] = "error" }
+    captives.each {| id | cohort_data[cohort][id.to_s.to_sym]["day#{cohort_day}"] = "captives" }
+    needs_id.each {| id | cohort_data[cohort][id.to_s.to_sym]["day#{cohort_day}"] = "needs_id" }
+    research.each {| id | cohort_data[cohort][id.to_s.to_sym]["day#{cohort_day}"] = "research" }
   end
+end
 
-  users_set[user.id] = {
-    user_id: user.id,
-    observation_id: observation.id,
-    latitude: geoip_response.results.ll[0],
-    longitude: geoip_response.results.ll[1],
-    city: geoip_response.results.city,
-    group: group,
-    errors: errors,
-    set: set
-  }
+# now sort out current day
+start_time = Time.now
+end_date = current_day
+ad = SegmentationStatistic.generate_segmentation_int( end_date - 1.day, end_date, use_database: true )
```

Comment:
Will this eventually get replaced when https://github.com/inaturalist/inaturalist/issues/4099 is implemented? I think this is OK, but FWIW I don't consider Kibana to be critical production infrastructure. We don't have the same uptime expectations, monitoring, or backups for it, so keep that in mind when deciding to query it

Reply:
this is unrelated to #4099 as it deals with first week retention - but the plan is to move it into the infrastructure if its successful at increasing retention or yielding useful data towards that goal.